### DISCUSSION TITLE
Web crawling plugin?

### DISCUSSION THREAD

**Discussion by Nikola-Milovic** [noib3/nvim-oxi#91] [Q&A]
Would it be possible to create something like a web crawling plugin? I read the examples, snooped around the issues and tried my hands at the project. This might be because I am not 100% familiar with Neovim API but I am not sure how to approach my idea. I would need a two-way communication between the Neovim and my plugin, also it would require statefulness as the browser would have to be persisted and kept in memory. But my problem is that rust doesn't have notion of global variables (for the better) and I cannot serialize the browser to keep it as state. My other solution would be to have a infinite loop function that would keep the browse in scope but since Neovim is single-threaded, it's probably not possible to block. Does anyone have some pointers or suggestions on where can I go from here or whether or not this is even possible?

**Comment by bsamseth** [comment]
You _can_ do global variables in Rust. Have a look at [once_cell](https://crates.io/crates/once_cell), specifically the `Lazy` type showcased in the second example. That works just fine and lets you keep state on the rust-side. I do this in a plugin I'm working on without issue. Global state might be generally discouraged, but there is a place for it, and this is one of them.
